presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java

```diff
@@ -0,0 +1,246 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.SystemSessionProperties;
+import com.facebook.presto.sql.tree.ArithmeticBinaryExpression;
+import com.facebook.presto.sql.tree.BooleanLiteral;
+import com.facebook.presto.sql.tree.Cast;
+import com.facebook.presto.sql.tree.ComparisonExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.Identifier;
+import com.facebook.presto.sql.tree.LambdaArgumentDeclaration;
+import com.facebook.presto.sql.tree.LambdaExpression;
+import com.facebook.presto.sql.tree.Literal;
+import com.facebook.presto.sql.tree.LongLiteral;
+import com.facebook.presto.sql.tree.NodeLocation;
+import com.facebook.presto.sql.tree.NotExpression;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.SearchedCaseExpression;
+import com.facebook.presto.sql.tree.WhenClause;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Optional;
+
+public class SimplifyArrayOperationsRewriter
+{
+    private SimplifyArrayOperationsRewriter() {}
+
+    public static Expression simplifyArrayOperations(Expression expression, Session session)
+    {
+        if (SystemSessionProperties.isSimplifyArrayOperations(session)) {
+            return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression);
+        }
+
+        return expression;
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private static boolean isConstant(Expression expression)
+        {
+            if (expression instanceof Cast) {
+                return isConstant(((Cast) expression).getExpression());
+            }
+
+            return expression instanceof Literal;
+        }
+
+        private static boolean isZero(Expression expression)
+        {
+            if (expression instanceof Cast) {
+                return isZero(((Cast) expression).getExpression());
+            }
+
+            return expression instanceof LongLiteral && ((LongLiteral) expression).getValue() == 0;
+        }
+
+        private static boolean isCardinalityOfFilter(Expression expression)
+        {
+            if (expression instanceof FunctionCall) {
+                FunctionCall functionCall = (FunctionCall) expression;
+                return functionCall.getName().toString().equals("cardinality") &&
+                        functionCall.getArguments().size() == 1 &&
```

Comment:
ditto to this. `cardinality` should always has single argument, if not, we should throw.

Reply:
I don't like throwing from the optimizer. This is just opportunistic optimization and be defensive and have the exact pattern that you want to match,

Reply:
Theoretically you don't need to check that here. Function resolution would have failed if arguments size is not 1. Agree with @hellium01's earlier point that this should be added to `RowExpressionRewriteRuleSet`. Also when you do that, please use proper API (add `isCardinalityFunction` `isFilterFunction` to `StandardFunctionResolution`) rather than relying on name (`CallExpression`'s name is only used for display purpose)